### PR TITLE
Write design patterns

### PR DESCRIPTION
**Pull request by cilim**
Task: [297](https://app.productive.io/1-infinum/tasks/1043925)

#### Aim

It would be nice to have a chapter that roughly describes what design patterns are and what are the most used ones in the Ruby world and that we use.

From a perspective of a developer I think that we'll use this chapter when we're faced with situations where we're planning a feature and aren't sure what the right approach is.
That's why we'll need a structured way to make this chapter helpful and easy to find what you need.

I skimmed through the **Design patterns in Ruby** book by _Russ Olsen_, read a couple of articles and went through all the design patterns on [Refactoring Guru](https://refactoring.guru/design-patterns) and the Ruby community is pretty synched in regards to the design pattern naming conventions (with the exception of different Factory patterns - we usually use the Simple Factory)

#### Solution

I summarized most of the Design patterns that can be found on [Refactoring Guru](https://refactoring.guru/design-patterns). It's a very nice site and all the examples are explained with great detail. There's also Ruby examples so I don't think we need to dig in much further than that, and our Handbook chapter will just be a quick reference to it, that helps developers find what they need (with the **Problem** sections).

I deployed the chapter to our [staging environment](https://beta.infinum.com/handbook/books/rails/development-practices/design-patterns).
Check it out and please help me decide on:
* should I continue giving a small example of the design pattern as I did in the first 2 chapters?
* did I miss any design patterns?
* should I include more specific design patterns that belong to families of the main patterns described in this chapter? (eg, **Null Object** pattern is just an improvement to error handling when dealing with the **Factory Method**)

### CODE REVIEW

**Comment by d4be4st** [comment]
omg i did not know about refactoring.guru website. looks awesome

My two cents:
* Would not include code examples, because then you would have to keep them updated. Would rather have different links with multiple examples.
* Some of those patterns i have never used like Prototype
* Different names than we usually use. What would be a Form object?
* Missing are some of the ones i most use: Service (or Command as guru calls it), State
* Rails specific patterns are missing like Repository, Form ...

**Comment by cilim** [comment]
Yeah, refactoring.guru is definitely something I'm planning on researching in depth :)

> Would not include code examples, because then you would have to keep them updated. Would rather have different links with multiple examples

that's a valid point, but since it's a design pattern, I thought there's little chance of us having to maintain it, cause I don't see it changing 🤔 

> Different names than we usually use

Maybe we should get on board with the community? :D 

> What would be a Form object?

I was thinking about that and I would place it in the Creational Design Pattern category, but if you try and search anything more on form objects outside the context of Rails, I can't find any results.
Since it makes sense to keep it because we use it on a daily basis, I'll find some article that's not from refactoring.guru and reference it.

> Missing are some of the ones i most use: Service (or Command as guru calls it), State

I'll add the `Command` and the Rails specific ones then (State [is here](https://github.com/infinum/rails-handbook/pull/117/files#diff-f5cea16a5f762555e1d1efbf4c751f96R192)).

**Comment by cilim** [comment]
@d4be4st all done:

- added table of contents
- added missing design patterns
- restructured and added references

Changes can be found [here](https://beta.infinum.com/handbook/books/rails/development-practices/design-patterns).

**Comment by d4be4st** [comment]
Can you try with details html tag to make it more readable?

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/details

**Comment by cilim** [comment]
@d4be4st what exactly do you want hidden?

**Comment by d4be4st** [comment]
The content. was thinking of just showing the titles which are expandable.

**Comment by cilim** [comment]
I don't think that's a good way to go, cause then you have to click on every design pattern to see what it solves and does it suit you and your problem :)

But that's just my opinion :) Maybe @infinum/backend-team can jump in too :) (hopefully everyone has Infinum VPN :D )

**Comment by cilim** [comment]
@melcha @d4be4st I had to extend the CSS a bit for the collapsing details to work.

The result can be seen in the chapter: https://beta.infinum.com/handbook/books/rails/development-practices/design-patterns

I also only applied it on the [#factory](https://beta.infinum.com/handbook/books/rails/development-practices/design-patterns#factory) chapters, cause it'll take some time to restructure all patterns in this way (everything inside `<details>` should be HTML, not MD anymore).
Let me know if this is the way we should go and I'll apply it to the rest.

**Comment by melcha** [comment]
@cilim this looks much more readable! Could you somehow highlight the summary content, so we know it's clickable?

**Comment by cilim** [comment]
@melcha something discreet

Screenshot 2020-06-16 at 15 46 22


or something more fun? 😄 

Screenshot 2020-06-16 at 15 45 42

File: Development practices/Design patterns.md [hunk 1]
```diff
@@ -1,12 +1,274 @@
...
+  <a href="https://refactoring.guru/design-patterns/factory-method/ruby/example">Ruby example</a>.
+</details>
+
+## Special object
+
+### Form object
+Simplify record operations by keeping logic in them.
+
+#### Problem
+* Your controllers are huge and you want to make them more readable by extracting some of its logic, but don't want to clutter your models.
```
Comment (cukoreva) [inline Development practices/Design patterns.md:83]:
```suggestion
* Your controllers are huge and you want to make them more readable by extracting some of their logic, but you don't want to clutter your models.
```

File: Development practices/Design patterns.md [hunk 2]
```diff
@@ -1,12 +1,274 @@
...
+* Your opinion is that models shouldn't have validations.
+* You have a virtual attribute that works only on a couple of places, but doesn't make sense to add `attr_accessor` to the model.
```
Comment (cukoreva) [inline Development practices/Design patterns.md:85]:
```suggestion
* You have a virtual attribute that works only in a couple of places, but it doesn't make sense to add `attr_accessor` to the model.
```

File: Development practices/Design patterns.md [hunk 3]
```diff
@@ -1,12 +1,274 @@
...
+* logic that is tightly coupled with primitives (attributes with behaviour - methods)
+
+#### Solution
+Read more about *Value object* [here](https://revs.runtime-revolution.com/value-objects-in-ruby-on-rails-9df64bc8db34).
+
+### Null object
+Instead of returning null, or some odd value, return a Special Case that has the same interface as what the caller expects.
+
+#### Problem
+You have a lot of conditionals scattered around many classes or view files that ask the same thing. It's usually a low-level error handling part that covers calls to `nil`s.
```
Comment (cukoreva) [inline Development practices/Design patterns.md:106]:
```suggestion
You have a lot of conditionals scattered around many classes or view files that check the same thing. It's usually a low-level error handling part that covers calls to `nil`s.
```

File: Development practices/Design patterns.md [hunk 4]
```diff
@@ -1,12 +1,274 @@
...
+* You have a complex SQL query that is used in multiple callers
+
+#### Solution
+Read more about *Query object* [here](https://medium.flatstack.com/query-object-in-ruby-on-rails-56ea434365f0).
+
+### Service object
+Concentrate the core logic of a request operation into a separate object.
+
+#### Problem
+You have a complex set of operations that need to be executed in order synchronously or asynchronously.
```
Comment (cukoreva) [inline Development practices/Design patterns.md:126]:
```suggestion
You have a complex set of operations that need to be executed in a sequence synchronously or asynchronously.
```

File: Development practices/Design patterns.md [hunk 5]
```diff
@@ -1,12 +1,274 @@
...
+
+#### Solution
+Read more about *Service object* [here](http://brewhouse.io/blog/2014/04/30/gourmet-service-objects.html).
+_Note_: Service objects can easily become a code smell if not handled properly. If you find yourself with a service object that has too many methods and/or becomes generally unreadable and hard to understand, you might have to find a [bounded context](https://blog.carbonfive.com/bring-clarity-to-your-monolith-with-bounded-contexts/) or use one of the design patterns mentioned in this chapter.
+
+### Builder
+Unlike the factories, this design pattern doesn't have a common interface. That makes it possible to different products, using the same construction process.
```
Comment (cukoreva) [inline Development practices/Design patterns.md:133]:
```suggestion
Unlike the factories, this design pattern doesn't have a common interface. That makes it possible to create different products, using the same construction process.
```

File: Development practices/Design patterns.md [hunk 6]
```diff
@@ -1,12 +1,274 @@
...
+Copy existing objects without making your code dependent on their classes.
+
+#### Problem
+You have an ActiveRecord object with lots of relationships that you want to clone.
+
+#### Solution
+Read more about *Prototype* [here](https://refactoring.guru/design-patterns/prototype).
+
+[Ruby example](https://refactoring.guru/design-patterns/prototype/ruby/example).
+_Note_: The Ruby example uses the usual Marshalling hack to make a deep copy. But it's rather slow and inefficient, therefore, in real applications, use a special gem (eg [deep_cloneable](https://github.com/moiristo/deep_cloneable))
```
Comment (cukoreva) [inline Development practices/Design patterns.md:153]:
```suggestion
_Note_: The Ruby example uses the usual Marshalling hack to make a deep copy. But it's rather slow and inefficient, therefore, in real applications, use a special gem (e.g. [deep_cloneable](https://github.com/moiristo/deep_cloneable))
```

File: Development practices/Design patterns.md [hunk 7]
```diff
@@ -1,12 +1,274 @@
...
+
+### Singleton
+Ensure a class to only have 1 object.
```
Comment (cukoreva) [inline Development practices/Design patterns.md:156]:
```suggestion
Ensure a class has only one object.
```

File: Development practices/Design patterns.md [hunk 8]
```diff
@@ -1,12 +1,274 @@
...
+#### Problem
+You need a way to transform your ActiveRecord object (or collection) into a valid XML request body.
+
+#### Solution
+Read more about *Adapter* [here](https://refactoring.guru/design-patterns/adapter).
+
+[Ruby example](https://refactoring.guru/design-patterns/adapter/ruby/example).
+
+### Bridge
+Split a large class or a set of closely related classes into two separate hierarchies
```
Comment (cukoreva) [inline Development practices/Design patterns.md:179]:
```suggestion
Split a large class or a set of closely related classes into two separate hierarchies.
```

Comment (cukoreva) [reply Development practices/Design patterns.md:179]: Please check if there are other places where the full stop is missing.

File: Development practices/Design patterns.md [hunk 9]
```diff
@@ -1,12 +1,274 @@
...
+
+#### Problem
+You're planning to add new subclasses which would complicate the class hierarchy.
```
Comment (cukoreva) [inline Development practices/Design patterns.md:182]:
```suggestion
You're planning to add new subclasses, which would complicate the class hierarchy.
```

File: Development practices/Design patterns.md [hunk 10]
```diff
@@ -1,12 +1,274 @@
...
+Read more about *Composite* [here](https://refactoring.guru/design-patterns/composite).
+
+[Ruby example](https://refactoring.guru/design-patterns/composite/ruby/example).
+
+### Decorator
+Add new behaviour to objects by placing them inside special wrapper objects.
+
+#### Problem
+You need a method that is using an ActiveRecord object attribute during the operation.
+You don't want to add the new method to the model cause it's only used in 1 place.
```
Comment (cukoreva) [inline Development practices/Design patterns.md:206]:
```suggestion
You don't want to add the new method to the model cause it's only used in one place.
```
Number 0-10 are normally written with words (when part of a text).

File: Development practices/Design patterns.md [hunk 11]
```diff
@@ -1,12 +1,274 @@
...
+
+#### Solution
+Read more about *Decorator* [here](https://refactoring.guru/design-patterns/decorator).
+
+[Ruby example](https://refactoring.guru/design-patterns/decorator/ruby/example).
+_Note_: The [draper](https://github.com/drapergem/draper) gem does exactly what you want.
+
+### Facade
+Provides a well documented, simplified and limited interface to a more complex system.
```
Comment (cukoreva) [inline Development practices/Design patterns.md:215]:
```suggestion
Provides a well-documented, simplified and limited interface to a more complex system.
```

File: Development practices/Design patterns.md [hunk 12]
```diff
@@ -1,12 +1,274 @@
...
+#### Solution
+Read more about *Strategy* [here](https://refactoring.guru/design-patterns/strategy).
+
+[Ruby example](https://refactoring.guru/design-patterns/strategy/ruby/example).
+
+### Template Method
+Define a skeleton of an algorithm in a base class and let subclasses override the steps without changing the overall algorithm’s structure.
+
+#### Problem
+You have many classes that solve the same semantical issue (importing and data processing from files, JSON/CSV/Excel) and have the same algorithm, but the parsing is different.
```
Comment (cukoreva) [inline Development practices/Design patterns.md:269]:
```suggestion
You have many classes that solve the same semantic issue (importing and data processing from files, JSON/CSV/Excel) and have the same algorithm, but the parsing is different.
```

File: Development practices/Design patterns.md [hunk 13]
```diff
@@ -1,12 +1,274 @@
...
+#### Problem
+Your application requirements changed and now you have to communicate with a 2nd API that returns same type of data as the 1st one.
+
+#### Solution
+Read more about *Strategy* [here](https://refactoring.guru/design-patterns/strategy).
+
+[Ruby example](https://refactoring.guru/design-patterns/strategy/ruby/example).
+
+### Template Method
+Define a skeleton of an algorithm in a base class and let subclasses override the steps without changing the overall algorithm’s structure.
```
Comment (cukoreva) [inline Development practices/Design patterns.md:266]:
```suggestion
Define a skeleton of an algorithm in a base class and let subclasses override the steps without changing the overall algorithm structure.
```

File: Development practices/Design patterns.md [hunk 14]
```diff
@@ -1,12 +1,274 @@
...
+#### Solution
+Read more about *State* [here](https://refactoring.guru/design-patterns/state).
+
+[Ruby example](https://refactoring.guru/design-patterns/state/ruby/example).
+
+### Strategy
+Turn a set of behaviours into objects and make them interchangeable inside the original context object.
+
+#### Problem
+Your application requirements changed and now you have to communicate with a 2nd API that returns same type of data as the 1st one.
```
Comment (cukoreva) [inline Development practices/Design patterns.md:258]:
```suggestion
Your application requirements changed and now you have to communicate with a second API that returns the same type of data as the first one.
```

File: Development practices/Design patterns.md [hunk 15]
```diff
@@ -1,12 +1,274 @@
...
+
+
+### Abstract factory
+One abstraction ladder step higher from the [Factory method](#factory-method) will get you to this design pattern. It lets you produce families of related objects without specifying their concrete classes.
+
+<details>
+  <summary>Solution</summary>
+  Read more about <b>Abstract factory</b> <a href="https://refactoring.guru/design-patterns/abstract-factory">here</a>.
+  <br>
+  <a href="https://refactoring.guru/design-patterns/factory-method/ruby/example">Ruby example</a>.
```
Comment (melcha) [inline Development practices/Design patterns.md:74]:
```suggestion
  <a href="https://refactoring.guru/design-patterns/abstract-factory/ruby/example">Ruby example</a>.
```

**Comment by melcha** [comment]
@cilim this looks much more readable! Could you somehow highlight the summary content, so we know it's clickable?

> @melcha something discreet
> 
> Screenshot 2020-06-16 at 15 46 22
> 
> or something more fun? 😄
> 
> Screenshot 2020-06-16 at 15 45 42

Let's go with the first one, but without the underline

**Comment by vr4b4c** [comment]
This is chapter is too broad with too many patterns covered. I would simplify it by dividing it into two sections. 

*Generic/standard/classic design patterns*
• general purpose of design patterns (why)
• giant link to code guru for more details about each pattern (how)

*Rails specific patterns*
• Form object (in conjunction with ActiveType)
• Query object
• Service object
• Decorator (Draper, SimpleDelegator, Forwardable)

**Comment by cilim** [comment]
@vr4b4c since the majority of people that reviewed the PR agree with style > readability (in context of hiding the `Problem` and `Solution` detail sections which I intended to simplify finding what you need) I believe that this is an ok argument to keep all general design patterns just behind a huge link to refactoring.guru, and keep the Rails specific patterns as they are.

What do you think @melcha @d4be4st ?

**Comment by melcha** [comment]
Actually I do agree with @vr4b4c. Initially I thought we would cover only Rails related patterns we use. 
There's no need to bloat the chapter with references to selected general purpose design patterns (we don't link all of them) if we can just link to refactoring guru, they've covered everything and the site looks great.

**Comment by cilim** [comment]
Changes applied with a fixup.

Deployed to staging. Check it out: https://beta.infinum.com/handbook/books/rails/development-practices/design-patterns

**Comment by vr4b4c** [comment]
Structurally, exactly on point with what I was suggesting. Stylewise, I don't think there is a need for collapsible Problem/Solution sections.

**Comment by cilim** [comment]
>Structurally, exactly on point with what I was suggesting. Stylewise, I don't think there is a need for collapsible Problem/Solution sections.

@d4be4st @melcha you were rooting for those, what do you say? Maybe it won't look bad now that there's much less content.

**Comment by melcha** [comment]
Only because there was a lot of content and no support for h4. We can remove it now.

**Comment by cilim** [comment]
All done.

File: Development practices/Design patterns.md [hunk 16]
```diff
@@ -1,12 +1,74 @@
...
+### Query object
+Represent a database query or a set of database queries related to the same table.
+
+#### Problem
+* You want to simplify a class that fetches some data from your database using a complex query
+* You want to simplify your model scope to use another class, instead of keeping the query in the caller
+* You have a complex SQL query that is used in multiple callers
+
+#### Solution
+Read more [here](https://medium.flatstack.com/query-object-in-ruby-on-rails-56ea434365f0)..
```
Comment (cukoreva) [inline Development practices/Design patterns.md:64]:
```suggestion
Read more [here](https://medium.flatstack.com/query-object-in-ruby-on-rails-56ea434365f0).
```

Comment (cilim) [reply Development practices/Design patterns.md]: Good 👁️ !

File: Development practices/Design patterns.md [hunk 17]
```diff
@@ -1,12 +1,74 @@
...
+### Value object
+A small simple object, like a date range or location, whose equality isn’t based on identity.
+
+#### Problem
+The problem can be spotted by finding:
+* attributes that don't make sense on their own
+* logic that is tightly coupled with primitives (attributes with behaviour - methods)
+
+#### Solution
+Read more [here](https://revs.runtime-revolution.com/value-objects-in-ruby-on-rails-9df64bc8db34).
```
Comment (d4be4st) [inline Development practices/Design patterns.md:44]: would mention https://dry-rb.org/gems/dry-initializer/3.0/ here

File: Development practices/Design patterns.md [hunk 18]
```diff
@@ -1,12 +1,74 @@
...
+They define a language that efficiently improves team communication.
+
+They can also be used to improve code readability and cleanness, as well as to simplify the design.
+
+[Refactoring.guru](https://refactoring.guru) is a great online place to learn more about them.
+
+## Rails patterns
+
+### Factory
+Class that has one creation method with a large conditional.
```
Comment (d4be4st) [inline Development practices/Design patterns.md:15]: A class that creates other classes? :D

File: Development practices/Design patterns.md [hunk 19]
```diff
@@ -1,12 +1,74 @@
...
+
+#### Problem
+You have a method which instantiates a different type of object depending on a particular parameter.
+
+#### Solution
+Read more [here](https://www.sihui.io/design-pattern-factory/).
+
+### Form object
+Simplify record operations by keeping logic in them.
```
Comment (d4be4st) [inline Development practices/Design patterns.md:24]: I would refraze this somehow...
The old definition was somehow better for me.

**Comment by cilim** [comment]
@infinum/backend-team any other volunteers eager to comment? :D Or planning to comment...

It's somewhat unproductive to revisit this branch after every N days when someone comments something new 🙉

**Comment by melcha** [comment]
You can merge it 👍

File: Development practices/Design patterns.md (hunk 18 above)
Comment (cilim) [reply Development practices/Design patterns.md:15]: Just saw this!

I like it somewhat more descriptive :D It technically doesn't only need to return a class, but can also return an object.

File: Development practices/Design patterns.md (hunk 19 above)
Comment (cilim) [reply Development practices/Design patterns.md]: @d4be4st last changes: https://github.com/infinum/rails-handbook/pull/117/commits/b2ed21578f5a003fc562f5f30fefc32b17318ee6#diff-f5cea16a5f762555e1d1efbf4c751f96R24

Let me know if you agree with @melcha and we can merge it.

**Pull request merged**